CascStorageLib/WDC3Reader.cs

```diff
@@ -0,0 +1,423 @@
+                        List<SparseEntry> tempSparseEntries = new List<SparseEntry>();
+                        for (int i = 0; i < (MaxIndex - MinIndex + 1); i++)
+                        {
+                            SparseEntry sparse = reader.Read<SparseEntry>();
+
+                            if (sparse.Offset == 0 || sparse.Size == 0)
+                                continue;
+
+                            // special case, may contain duplicates in the offset map that we don't want
+                            if (sections[sectionIndex].CopyTableCount == 0)
+                            {
+                                if (offSetKeyMap.ContainsKey(sparse.Offset))
+                                    continue;
+                            }
+
+                            tempSparseEntries.Add(sparse);
+                            offSetKeyMap.Add(sparse.Offset, 0);
+                        }
+
+                        sparseEntries = tempSparseEntries.ToArray();
+                    }
+
+                    // index data
+                    m_indexData = reader.ReadArray<int>(sections[sectionIndex].IndexDataSize / 4);
+
+                    // duplicate rows data
+                    Dictionary<int, int> copyData = new Dictionary<int, int>();
+
+                    for (int i = 0; i < sections[sectionIndex].CopyTableCount; i++)
+                        copyData[reader.ReadInt32()] = reader.ReadInt32();
+
+                    // reference data
+                    ReferenceData refData = null;
+
+                    if (sections[sectionIndex].ParentLookupDataSize > 0)
+                    {
+                        refData = new ReferenceData
+                        {
+                            NumRecords = reader.ReadInt32(),
+                            MinId = reader.ReadInt32(),
+                            MaxId = reader.ReadInt32()
+                        };
+
+                        refData.Entries = reader.ReadArray<ReferenceEntry>(refData.NumRecords);
+                    }
+
+                    int position = 0;
+
+                    bool indexDataNotEmpty = sections[sectionIndex].IndexDataSize != 0 && m_indexData.GroupBy(i => i).Where(d => d.Count() > 1).Count() == 0;
+
+                    for (int i = 0; i < RecordsCount; ++i)
+                    {
+                        BitReader bitReader = new BitReader(recordsData) { Position = 0 };
+
+                        if (Flags.HasFlagExt(DB2Flags.Sparse))
+                        {
+                            bitReader.Position = position;
+                            position += sparseEntries[i].Size * 8;
+                        }
+                        else
+                            bitReader.Offset = i * RecordSize;
+
+                        IDB2Row rec = new WDC2Row(this, bitReader, sections[sectionIndex].FileOffset, indexDataNotEmpty ? m_indexData[i] : -1, refData?.Entries[i], i);
```

Comment:
WDC3Row ?

Reply:
Oh, this PR is still open, oops. Yeah don't nitpick too much on that, when I saw it worked I shipped it and went to bed.